### PR TITLE
HDDS-2159. Fix Race condition in ProfileServlet#pid.

### PR DESCRIPTION
**Pull request by hanishakoneru**
There is a race condition in ProfileServlet. The Servlet member field pid should not be used for local assignment. It could lead to race condition.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 36 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| -1 | mvninstall | 28 | hadoop-ozone in trunk failed. |
| -1 | compile | 22 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 51 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| -1 | shadedclient | 99 | branch has errors when building and testing our client artifacts. |
| +1 | javadoc | 144 | trunk passed |
| 0 | spotbugs | 158 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 25 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| -1 | mvninstall | 30 | hadoop-ozone in the patch failed. |
| -1 | compile | 21 | hadoop-ozone in the patch failed. |
| -1 | javac | 21 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 50 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| -1 | shadedclient | 31 | patch has errors when building and testing our client artifacts. |
| +1 | javadoc | 146 | the patch passed |
| -1 | findbugs | 23 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 228 | hadoop-hdds in the patch failed. |
| -1 | unit | 26 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 1687 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.container.placement.algorithms.TestSCMContainerPlacementRackAware |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1488 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e5ff34138d70 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3f223be |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-compile-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-compile-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/testReport/ |
| Max. process+thread count | 487 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/framework U: hadoop-hdds/framework |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ProfileServlet.java [hunk 1]
```diff
@@ -208,11 +208,11 @@ protected void doGet(final HttpServletRequest req,
       return;
     }
     // if pid is explicitly specified, use it else default to current process
-    pid = getInteger(req, "pid", pid);
+    Integer processId = getInteger(req, "pid", pid);
 
     // if pid is not specified in query param and if current process pid
```
Comment (xiaoyuyao) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ProfileServlet.java]: should we update the comment? It seems we are not trying to determine current process pid?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 1 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| -1 | mvninstall | 31 | hadoop-ozone in trunk failed. |
| -1 | compile | 22 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 50 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| -1 | shadedclient | 100 | branch has errors when building and testing our client artifacts. |
| -1 | javadoc | 47 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 186 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 25 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| -1 | mvninstall | 29 | hadoop-ozone in the patch failed. |
| -1 | compile | 20 | hadoop-ozone in the patch failed. |
| -1 | javac | 20 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 51 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| -1 | shadedclient | 32 | patch has errors when building and testing our client artifacts. |
| -1 | javadoc | 47 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 22 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| +1 | unit | 243 | hadoop-hdds in the patch passed. |
| -1 | unit | 26 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 1714 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1488 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 250fc8e627d1 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d7d6ec8 |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/testReport/ |
| Max. process+thread count | 487 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/framework U: hadoop-hdds/framework |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1488/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ProfileServlet.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ProfileServlet.java]: It is determined in `getPid()` call from constructor:

https://github.com/apache/hadoop/blob/aa664d72595ddfcb1a1bf082381bb222e59db354/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ProfileServlet.java#L126-L128

**Comment by anuengineer** [comment]
I ran the checkstyle 3 times and confirmed that at this point of time this patch is clean of issues.
@adoroszlai  and @xiaoyuyao  Thanks for the reviews. @hanishakoneru Thank you for the contribution. I have committed this patch to the trunk.

**Pull request closed without merging**